tasmota/tasmota_xsns_sensor/xsns_33_qmc5883l.ino

```diff
@@ -0,0 +1,331 @@
+    return; // Software Reset
+  if (I2cWrite8(QMC5883L_ADDR, QMC5883L_RESET, 0x01) == false)
+    return;
+  // write config
+  if (I2cWrite8(QMC5883L_ADDR, QMC5883L_CONFIG, QMC5883L_CONFIG_OS256 | QMC5883L_CONFIG_8GAUSS | QMC5883L_CONFIG_100HZ | QMC5883L_CONFIG_CONT) == false)
+    return;
+  QMC5883L->ready = true;
+}
+
+//Read the magnetic data
+void QMC5883L_read_data(void)
+{
+  // check if chip is ready to provice data
+  if (!(I2cRead8(QMC5883L_ADDR, QMC5883L_STATUS) & QMC5883L_STATUS_DRDY))
+    return; // chip not yet ready, next round try again
+
+  QMC5883L->MX = I2cReadS16_LE(QMC5883L_ADDR, QMC5883L_X_LSB); // select LSB register
+  QMC5883L->MY = I2cReadS16_LE(QMC5883L_ADDR, QMC5883L_Y_LSB);
+  QMC5883L->MZ = I2cReadS16_LE(QMC5883L_ADDR, QMC5883L_Z_LSB);
+
+  // calculate scalar magnetic induction
+  QMC5883L->scalar = sqrt((QMC5883L->MX * QMC5883L->MX) + (QMC5883L->MY * QMC5883L->MY) + (QMC5883L->MZ * QMC5883L->MZ));
+
+  // get temperature
+  QMC5883L->temp = (I2cReadS16_LE(QMC5883L_ADDR, QMC5883L_TEMP_LSB) / 100) + QMC5883L_TEMP_SHIFT;
+}
+
+/*********************************************************************************************\
+ * Presentation
+\*********************************************************************************************/
+
+#ifdef USE_WEBSERVER
+const char HTTP_SNS_QMC5883L[] PROGMEM =
+    "{s}QMC5883L " D_MX "{m}%d " D_UNIT_MICROTESLA "{e}"                        // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+    "{s}QMC5883L " D_MY "{m}%d " D_UNIT_MICROTESLA "{e}"                        // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+    "{s}QMC5883L " D_MZ "{m}%d " D_UNIT_MICROTESLA "{e}"                        // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+    "{s}QMC5883L " D_MAGNETICFLD "{m}%d " D_UNIT_MICROTESLA "{e}"               // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+    "{s}QMC5883L " D_TEMPERATURE "{m}%d " D_UNIT_DEGREE D_UNIT_CELSIUS "{e}"; // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+
+const char HTTP_SNS_QMC5883L_ERROR[] PROGMEM =
+    "{s}QMC5883L {m} %s {e}";
+#endif
+
+
+void QMC5883L_Show(uint8_t json)
+{
+  if (json) 
+  {
+    ResponseAppend_P(PSTR(",\"QMC5883L\":{\"" D_JSON_MX "\":%d,\"" D_JSON_MY "\":%d,\"" D_JSON_MZ "\":%d,\"" D_JSON_MAGNETICFLD "\":%u,\"" D_JSON_TEMPERATURE "\":%d}"), QMC5883L->MX, QMC5883L->MY, QMC5883L->MZ, QMC5883L->scalar, QMC5883L->temp);
+  }
+#ifdef USE_WEBSERVER
+  else 
+  {
+    WSContentSend_PD(HTTP_SNS_QMC5883L, QMC5883L->MX, QMC5883L->MY, QMC5883L->MZ, QMC5883L->scalar, QMC5883L->temp);
+  }
+#endif
+}
+
+/*********************************************************************************************\
+ * Interface
+\*********************************************************************************************/
+ 
+bool Xsns33(byte function)
+{
+  if (!I2cEnabled(XI2C_71))
+  {
+    return false;
+  }
+  if (FUNC_INIT == function)
+  {
+    QMC5883L_Init();
+  }
+  else if (QMC5883L->ready == true)
```

Comment:
Wrong
I suggested a test : `if (QMC5883L)`
With your implementation if you dereference QMC5883L as a nullptr you read random value for `ready`
You MUST test QMC5883L as a non nullptr and this is what my suggestion is about from hte beginning